// @collaborne/media-player/dist/components/draggable-popover/useDraggablePopoverHook.js
import { useEffect, useRef, useState } from 'react';
import { useMeasure, usePrevious } from 'react-use';
import { DEFAULT_PIP_SIZE } from '../../utils/constants';
const vw = window.innerWidth;
const DEFAULT_DIMENSIONS = {
    width: '100%',
    height: '100%',
    y: 0,
    x: 0,
};
export const useDraggablePopoverHook = ({ isPip, xAxisDistance, yAxisDistance, pipDraggableAreaRef, }) => {
    // DraggablePopover DnD params:
    const [portalWrapperRef, containerSize] = useMeasure();
    // Track for PIP window(was it resized or moved before)
    const hasPipMovedOrResizedRef = useRef(false);
    // Track for window(was it resized or not) when PIP mode was active
    const hasWindowResizedRef = useRef(false);
    const prevContainerSize = usePrevious(containerSize);
    const firstPositionRef = useRef({
        width: DEFAULT_PIP_SIZE.width,
        height: DEFAULT_PIP_SIZE.height,
        x: 0,
        y: 0,
    });
    // Get player width + margin on 2 sides
    const pipPlayerWidth = DEFAULT_PIP_SIZE.width + xAxisDistance;
    const pipPlayerHight = DEFAULT_PIP_SIZE.height + yAxisDistance;
    const [dimensions, setDimensions] = useState(DEFAULT_DIMENSIONS);
    const enableResizing = !isPip
        ? false
        : {
            topLeft: true,
            topRight: true,
            bottomLeft: true,
            bottomRight: true,
        };
    const handleDragStop = (_e, d) => {
        hasPipMovedOrResizedRef.current = true;
        setDimensions(prev => ({ ...prev, x: d.x, y: d.y }));
    };
    const handleResizeStop = (_e, _direction, ref, _delta, position) => {
        hasPipMovedOrResizedRef.current = true;
        setDimensions(prev => ({
            ...prev,
            width: ref.style.width,
            height: ref.style.height,
            ...position,
        }));
    };
    // Track window resize events
    useEffect(() => {
        const onWindowResize = () => {
            if (!isPip) {
                hasWindowResizedRef.current = false;
                return;
            }
            hasWindowResizedRef.current = true;
        };
        window.addEventListener('resize', onWindowResize);
        return () => window.removeEventListener('resize', onWindowResize);
    }, [isPip]);
    // Get first mounting positions for the PIP player
    useEffect(() => {
        if (!pipDraggableAreaRef.current || hasPipMovedOrResizedRef.current) {
            return;
        }
        const width = pipDraggableAreaRef.current.offsetWidth;
        const height = pipDraggableAreaRef.current.offsetHeight;
        firstPositionRef.current = {
            x: width - pipPlayerWidth,
            y: height - pipPlayerHight,
            width: DEFAULT_PIP_SIZE.width,
            height: DEFAULT_PIP_SIZE.height,
        };
        // containerSize as a dependency means that we always will have right initial dimensions(user can resize browser on any stage)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [containerSize]);
    // Set first mounting position when PIP was switched
    useEffect(() => {
        // If: media is in PIP mode and wasn't resized or moved
        // then position it with initial values
        if (isPip && !hasPipMovedOrResizedRef.current) {
            return setDimensions(firstPositionRef.current);
        }
        if (!isPip) {
            hasPipMovedOrResizedRef.current = false;
            setDimensions(DEFAULT_DIMENSIONS);
        }
    }, [isPip]);
    // Calculates positions and dimensions
    useEffect(() => {
        var _a, _b;
        const hasOnlyWindowResized = hasWindowResizedRef.current && !hasPipMovedOrResizedRef.current;
        if (isPip && hasOnlyWindowResized) {
            return setDimensions(firstPositionRef.current);
        }
        if (isPip && hasPipMovedOrResizedRef.current) {
            const width = (_a = containerSize === null || containerSize === void 0 ? void 0 : containerSize.width) !== null && _a !== void 0 ? _a : document.body.offsetWidth;
            const prevWidth = (_b = prevContainerSize === null || prevContainerSize === void 0 ? void 0 : prevContainerSize.width) !== null && _b !== void 0 ? _b : vw;
            setDimensions(prev => {
                const newX = prev.x + width - prevWidth;
                return {
                    ...prev,
                    x: newX > 0 ? newX : 0,
                };
            });
        }
        // Should occur only in PIP mode and when is browser resized(windowSize changed)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isPip, containerSize]);
    return {
        enableResizing,
        dimensions,
        handleDragStop,
        handleResizeStop,
        portalWrapperRef,
    };
};
